iroh-relay/src/dns.rs

```diff
@@ -1,22 +1,259 @@
+//! DNS resolver
+
 use std::{
-    net::{IpAddr, Ipv6Addr},
-    sync::OnceLock,
+    fmt::Write,
+    future::Future,
+    net::{IpAddr, Ipv6Addr, SocketAddr},
+};
+
+use anyhow::{bail, Context, Result};
+use hickory_resolver::{lookup::TxtLookup, IntoName, ResolveError, Resolver, TokioResolver};
+use iroh_base::{NodeAddr, NodeId};
+use n0_future::{
+    time::{self, Duration},
+    StreamExt,
 };
+use url::Url;
 
-use anyhow::Result;
-use hickory_resolver::{Resolver, TokioResolver};
+pub mod node_info;
 
-/// The DNS resolver type used throughout `iroh`.
-pub(crate) type DnsResolver = TokioResolver;
+/// The n0 testing DNS node origin, for production.
+pub const N0_DNS_NODE_ORIGIN_PROD: &str = "dns.iroh.link";
+/// The n0 testing DNS node origin, for testing.
+pub const N0_DNS_NODE_ORIGIN_STAGING: &str = "staging-dns.iroh.link";
 
-static DNS_RESOLVER: OnceLock<TokioResolver> = OnceLock::new();
+/// The DNS resolver used throughout `iroh`.
+#[derive(Debug, Clone)]
+pub struct DnsResolver(TokioResolver);
 
-/// Get a reference to the default DNS resolver.
-///
-/// The default resolver can be cheaply cloned and is shared throughout the running process.
-/// It is configured to use the system's DNS configuration.
-pub fn default_resolver() -> &'static DnsResolver {
-    DNS_RESOLVER.get_or_init(|| create_default_resolver().expect("unable to create DNS resolver"))
+impl DnsResolver {
+    /// Create a new DNS resolver with sensible cross-platform defaults.
+    ///
+    /// We first try to read the system's resolver from `/etc/resolv.conf`.
+    /// This does not work at least on some Androids, therefore we fallback
+    /// to the default `ResolverConfig` which uses eg. to google's `8.8.8.8` or `8.8.4.4`.
+    pub fn new_with_defaults() -> Self {
```

Comment:
This is an odd name for rust conventions, why not just `new`?  And often `Default::default` is aliased to this as well, clippy even has a lint for when you impl new without args but not Default IIRC.

Reply:
Yeah - I see where you're coming from, but I often see `Default::default()` interpreted as "empty".
And in this case, the `DnsResolver` is sort of a container that "has" nameservers.
So I understand being hesitant implementing `Default::default` or just calling this `new`.

It *might* make sense to implement a `DnsResolver::builder()` that allows you to `.with_nameserver`, or `.with_defaults`, but in the end it might also mostly blow up use sites that now are mostly limited to a single call to `new_with_defaults` or `with_single_nameserver`.

Reply:
Ah - I wrote this up *right* before the latest two commits :upside_down_face: 
Whatever: I like the changed version, too!

Reply:
:+1: changed